Unify semantics of `cpblk` and `initblk` of zero size between Mono and CoreCLR

It was discovered in a CI run (`Mono monointerpreter Pri0 Runtime Tests Run OSX x64 release`) that (at least when using the interpreter) Mono throws a `NullReferenceException` when executing a `cpblk` with a size of zero (I suspect it is the same for `initblk`).

Reproduction:
```cs
using System;
using System.Runtime.CompilerServices;

Console.WriteLine(TestCpBlk(ref Unsafe.NullRef<byte>(), ref Unsafe.NullRef<byte>(), 0));

[MethodImpl(MethodImplOptions.NoInlining)]
static bool TestCpBlk(ref byte dst, ref byte src, uint size)
{
    Unsafe.CopyBlock(ref dst, ref src, size);

    return Unsafe.AreSame(ref dst, ref Unsafe.NullRef<byte>());
}
```

In ECMA 335, `III.3.30 cpblk`/`III.3.36 initblk`, we can find under `Exceptions`:
```
System.NullReferenceException can be thrown if an invalid address is detected.
```
Suggesting that both throwing and not throwing are legal behaviors (and they both seem reasonable). But it would be nice to have the two runtimes behave the same.

> I think we should fix Mono to stop throwing for null pointer and zero size for these opcodes.

As far as I can tell, this is just a Mono interpreter issue.  The JIT (and AOT) does not throw in this case (verified by inspecting method-to-ir.c and also by running a repro .csproj, below).

```csproj
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net6.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <RuntimeIdentifier>osx-x64</RuntimeIdentifier>
    <SelfContained>true</SelfContained>
    <UseMonoRuntime>true</UseMonoRuntime>
  </PropertyGroup>

</Project>
```

test with
```console
$ dotnet run
True
$ MONO_ENV_OPTIONS=--interp dotnet run

Unhandled Exception:
System.NullReferenceException: Object reference not set to an instance of an object.
   at Program.<<Main>$>g__TestCpBlk|0_0(Byte& dst, Byte& src, UInt32 size) in /private/tmp/CpblkTest/Program.cs:line 9
   at Program.<Main>$(String[] args) in /private/tmp/CpblkTest/Program.cs:line 4
[ERROR] FATAL UNHANDLED EXCEPTION: System.NullReferenceException: Object reference not set to an instance of an object.
   at Program.<<Main>$>g__TestCpBlk|0_0(Byte& dst, Byte& src, UInt32 size) in /private/tmp/CpblkTest/Program.cs:line 9
   at Program.<Main>$(String[] args) in /private/tmp/CpblkTest/Program.cs:line 4
```

/cc @BrzVlad

--- a/src/libraries/System.Private.CoreLib/src/System/Span.cs
+++ b/src/libraries/System.Private.CoreLib/src/System/Span.cs
@@ -284,18 +284,11 @@ namespace System
         {
             if (Unsafe.SizeOf<T>() == 1)
             {
-#if MONO
-                // Mono runtime's implementation of initblk performs a null check on the address.
-                // We'll perform a length check here to avoid passing a null address in the empty span case.
-                if (_length != 0)
-#endif
-                {
-                    // Special-case single-byte types like byte / sbyte / bool.
-                    // The runtime eventually calls memset, which can efficiently support large buffers.
-                    // We don't need to check IsReferenceOrContainsReferences because no references
-                    // can ever be stored in types this small.
-                    Unsafe.InitBlockUnaligned(ref Unsafe.As<T, byte>(ref _pointer.Value), Unsafe.As<T, byte>(ref value), (uint)_length);
-                }
+                // Special-case single-byte types like byte / sbyte / bool.
+                // The runtime eventually calls memset, which can efficiently support large buffers.
+                // We don't need to check IsReferenceOrContainsReferences because no references
+                // can ever be stored in types this small.
+                Unsafe.InitBlockUnaligned(ref Unsafe.As<T, byte>(ref _pointer.Value), Unsafe.As<T, byte>(ref value), (uint)_length);
             }
             else
             {

--- a/src/mono/mono/mini/interp/interp.c
+++ b/src/mono/mono/mini/interp/interp.c
@@ -6804,7 +6804,7 @@ MINT_IN_CASE(MINT_BRTRUE_I8_SP) ZEROP_SP(gint64, !=); MINT_IN_BREAK;
 		MINT_IN_CASE(MINT_CPBLK) {
 			gpointer dest = LOCAL_VAR (ip [1], gpointer);
 			gpointer src = LOCAL_VAR (ip [2], gpointer);
-			gint32 size = LOCAL_VAR (ip [3], gint32);
+			guint32 size = LOCAL_VAR (ip [3], guint32);
 			if (size && (!dest || !src))
 				THROW_EX (mono_get_exception_null_reference(), ip);
 			else
@@ -6814,7 +6814,7 @@ MINT_IN_CASE(MINT_BRTRUE_I8_SP) ZEROP_SP(gint64, !=); MINT_IN_BREAK;
 		}
 		MINT_IN_CASE(MINT_INITBLK) {
 			gpointer dest = LOCAL_VAR (ip [1], gpointer);
-			gint32 size = LOCAL_VAR (ip [3], gint32);
+			guint32 size = LOCAL_VAR (ip [3], guint32);
 			if (size)
 				NULL_CHECK (dest);
 			memset (dest, LOCAL_VAR (ip [2], gint32), size);

--- a/src/mono/mono/mini/interp/interp.c
+++ b/src/mono/mono/mini/interp/interp.c
@@ -6804,18 +6804,20 @@ MINT_IN_CASE(MINT_BRTRUE_I8_SP) ZEROP_SP(gint64, !=); MINT_IN_BREAK;
 		MINT_IN_CASE(MINT_CPBLK) {
 			gpointer dest = LOCAL_VAR (ip [1], gpointer);
 			gpointer src = LOCAL_VAR (ip [2], gpointer);
-			if (!dest || !src)
+			gint32 size = LOCAL_VAR (ip [3], gint32);
+			if (size && (!dest || !src))
 				THROW_EX (mono_get_exception_null_reference(), ip);
-			/* FIXME: value and size may be int64... */
-			memcpy (dest, src, LOCAL_VAR (ip [3], gint32));
+			else
+				memcpy (dest, src, size);
 			ip += 4;
 			MINT_IN_BREAK;
 		}
 		MINT_IN_CASE(MINT_INITBLK) {
 			gpointer dest = LOCAL_VAR (ip [1], gpointer);
-			NULL_CHECK (dest);
-			/* FIXME: value and size may be int64... */
-			memset (dest, LOCAL_VAR (ip [2], gint32), LOCAL_VAR (ip [3], gint32));
+			gint32 size = LOCAL_VAR (ip [3], gint32);
+			if (size)
+				NULL_CHECK (dest);
+			memset (dest, LOCAL_VAR (ip [2], gint32), size);
 			ip += 4;
 			MINT_IN_BREAK;
 		}

